<h1>Netflix Data</h1>
<p>In the age of pandemic, Netflix becomes an essential part of life, Kaggle user <b>Shivam Bansal</b> prepared a <a href="https://www.kaggle.com/shivamb/netflix-shows/code">dataset</a> including the list of movies and TV shows included in Netflix. Let's have a look of what they have for users</p>

<p>First, let's import the essential libraries and the dataset:</p>

In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('dataset/netflix_titles.csv')

<h2>Exploratory</h2>
<p>Let's have a look at the data, understand how it is structured, and what data does it include:</p>

In [3]:
df.head(5)

show_id     type  title           director  \
0      s1  TV Show     3%                NaN   
1      s2    Movie   7:19  Jorge Michel Grau   
2      s3    Movie  23:59       Gilbert Chan   
3      s4    Movie      9        Shane Acker   
4      s5    Movie     21     Robert Luketic   

                                                cast        country  \
0  João Miguel, Bianca Comparato, Michel Gomes, R...         Brazil   
1  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
2  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
3  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
4  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   

          date_added  release_year rating   duration  \
0    August 14, 2020          2020  TV-MA  4 Seasons   
1  December 23, 2016          2016  TV-MA     93 min   
2  December 20, 2018          2011      R     78 min   
3  November 16, 2017          2009  PG-13     80 min   
4    January 1, 2020          2008  PG-13    123 min   

                                           listed_in  \
0  International TV Shows, TV Dramas, TV Sci-Fi &...   
1                       Dramas, International Movies   
2                Horror Movies, International Movies   
3  Action & Adventure, Independent Movies, Sci-Fi...   
4                                             Dramas   

                                         description  
0  In a future where the elite inhabit an island ...  
1  After a devastating earthquake hits Mexico Cit...  
2  When an army recruit is found dead, his fellow...  
3  In a postapocalyptic world, rag-doll robots hi...  
4  A brilliant group of students become card-coun...

In [4]:
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [5]:
df['type'].value_counts()

Movie      5377
TV Show    2410
Name: type, dtype: int64

<h3>Movie Data</h3>
<p>Here we will be only foucsing on movie data only, so I would filter out the dataset.</p>

In [6]:
movie_df = df[df['type'] == 'Movie'].copy()

<p>As we have are focusing on movie data only, we could drop the 'type' column.</p>

In [7]:
movie_df.drop(columns = 'type', axis = 1, inplace = True)

In [8]:
movie_df

show_id                                    title           director  \
1         s2                                     7:19  Jorge Michel Grau   
2         s3                                    23:59       Gilbert Chan   
3         s4                                        9        Shane Acker   
4         s5                                       21     Robert Luketic   
6         s7                                      122    Yasir Al Yasiri   
...      ...                                      ...                ...   
7781   s7782                                     Zoom       Peter Hewitt   
7782   s7783                                     Zozo        Josef Fares   
7783   s7784                                   Zubaan        Mozez Singh   
7784   s7785                        Zulu Man in Japan                NaN   
7786   s7787  ZZ TOP: THAT LITTLE OL' BAND FROM TEXAS           Sam Dunn   

                                                   cast  \
1     Demián Bichir, Héctor Bonilla, Oscar Serrano, ...   
2     Tedd Chan, Stella Chung, Henley Hii, Lawrence ...   
3     Elijah Wood, John C. Reilly, Jennifer Connelly...   
4     Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...   
6     Amina Khalil, Ahmed Dawood, Tarek Lotfy, Ahmed...   
...                                                 ...   
7781  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
7782  Imad Creidi, Antoinette Turk, Elias Gergi, Car...   
7783  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   
7784                                            Nasty C   
7786                                                NaN   

                                                country          date_added  \
1                                                Mexico   December 23, 2016   
2                                             Singapore   December 20, 2018   
3                                         United States   November 16, 2017   
4                                         United States     January 1, 2020   
6                                                 Egypt        June 1, 2020   
...                                                 ...                 ...   
7781                                      United States    January 11, 2020   
7782  Sweden, Czech Republic, United Kingdom, Denmar...    October 19, 2020   
7783                                              India       March 2, 2019   
7784                                                NaN  September 25, 2020   
7786              United Kingdom, Canada, United States       March 1, 2020   

      release_year rating duration  \
1             2016  TV-MA   93 min   
2             2011      R   78 min   
3             2009  PG-13   80 min   
4             2008  PG-13  123 min   
6             2019  TV-MA   95 min   
...            ...    ...      ...   
7781          2006     PG   88 min   
7782          2005  TV-MA   99 min   
7783          2015  TV-14  111 min   
7784          2019  TV-MA   44 min   
7786          2019  TV-MA   90 min   

                                              listed_in  \
1                          Dramas, International Movies   
2                   Horror Movies, International Movies   
3     Action & Adventure, Independent Movies, Sci-Fi...   
4                                                Dramas   
6                   Horror Movies, International Movies   
...                                                 ...   
7781                 Children & Family Movies, Comedies   
7782                       Dramas, International Movies   
7783     Dramas, International Movies, Music & Musicals   
7784  Documentaries, International Movies, Music & M...   
7786                    Documentaries, Music & Musicals   

                                            description  
1     After a devastating earthquake hits Mexico Cit...  
2     When an army recruit is found dead, his fellow...  
3     In a postapocalyptic world, rag-doll robots hi...  
4     A brilliant group of students 

<h3>Missing Data</h3>
<p>The next step is to check if we have any null data, and as we could see there are some missing data for 'director', 'cast', 'country' and 'rating'.</p>

<p>Let's begin with rating, as it only has 5 missing data.</p>

In [9]:
print(movie_df.isnull().sum())
before = movie_df.isnull().sum()

show_id           0
title             0
director        163
cast            426
country         230
date_added        0
release_year      0
rating            5
duration          0
listed_in         0
description       0
dtype: int64


In [10]:
movie_df[movie_df['rating'].isnull()]

show_id                                              title  \
67       s68  13TH: A Conversation with Oprah Winfrey & Ava ...   
3736   s3737                                    Louis C.K. 2017   
3737   s3738                              Louis C.K.: Hilarious   
3738   s3739               Louis C.K.: Live at the Comedy Store   
4323   s4324                               My Honor Was Loyalty   

             director                                               cast  \
67                NaN                        Oprah Winfrey, Ava DuVernay   
3736       Louis C.K.                                         Louis C.K.   
3737       Louis C.K.                                         Louis C.K.   
3738       Louis C.K.                                         Louis C.K.   
4323  Alessandro Pepe  Leone Frisa, Paolo Vaccarino, Francesco Miglio...   

            country          date_added  release_year rating duration  \
67              NaN    January 26, 2017          2017    NaN   37 min   
3736  United States       April 4, 2017          2017    NaN   74 min   
3737  United States  September 16, 2016          2010    NaN   84 min   
3738  United States     August 15, 2016          2015    NaN   66 min   
4323          Italy       March 1, 2017          2015    NaN  115 min   

     listed_in                                        description  
67      Movies  Oprah Winfrey sits down with director Ava DuVe...  
3736    Movies  Louis C.K. muses on religion, eternal love, gi...  
3737    Movies  Emmy-winning comedy writer Louis C.K. brings h...  
3738    Movies  The comic puts his trademark hilarious/thought...  
4323    Dramas  Amid the chaos and horror of World War II, a c...

<p>With only five missing data, we coul deasily deal with them individual by doing some manual search. Another interesting finding is that the dataset has included "comedy show" as movie, which is quite debatable. Anyway, we could leave them for now.</p>

<p>Speaking of which, we could also drop the <b>"show_id"</b> column as it does not really tell us much.</p>

In [11]:
movie_df.drop(columns = 'show_id', axis = 1, inplace = True)

In [12]:
movie_df.head(4)

title           director  \
1   7:19  Jorge Michel Grau   
2  23:59       Gilbert Chan   
3      9        Shane Acker   
4     21     Robert Luketic   

                                                cast        country  \
1  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
2  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
3  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
4  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   

          date_added  release_year rating duration  \
1  December 23, 2016          2016  TV-MA   93 min   
2  December 20, 2018          2011      R   78 min   
3  November 16, 2017          2009  PG-13   80 min   
4    January 1, 2020          2008  PG-13  123 min   

                                           listed_in  \
1                       Dramas, International Movies   
2                Horror Movies, International Movies   
3  Action & Adventure, Independent Movies, Sci-Fi...   
4                                             Dramas   

                                         description  
1  After a devastating earthquake hits Mexico Cit...  
2  When an army recruit is found dead, his fellow...  
3  In a postapocalyptic world, rag-doll robots hi...  
4  A brilliant group of students become card-coun...

In [13]:
movie_df.reset_index(inplace = True)

<p>Missing data of movies should be relatively easy to resolve, here, I would make use of the <a href="https://www.themoviedb.org/">TMDB (The Movie Database) API</a> to fill in the missing data.</p>
<p><em>Oh, and we also need to make sure we have installed and imported the 'requests' package.</em></p>
<p>So I have create 4 simple functions to deal with this problem (please use your own API keys, it is free.).</p>

<h4>Getting Movie ID</h4>
<p>To fill in the missing data, we would need to first create a new column, and get the movie id for each movie.</p>

In [14]:
import requests
from configs import TMDB_API

In [15]:
# Find Movie ID in TMDB
def get_movie_id(movie_name):
    response = requests.get(
    'https://api.themoviedb.org/3/search/movie',
    params={'api_key': TMDB_API,
            'query' : movie_name},
    )
    
    try:
        movie_id = response.json()['results'][0]['id']     
#       print('Movie ID found: ', movie_id)
    except:
        movie_id = np.nan
      
    return movie_id

In [16]:
movie_df['movie_id'] = np.nan

In [17]:
movie_df[:10]

index  title             director  \
0      1   7:19    Jorge Michel Grau   
1      2  23:59         Gilbert Chan   
2      3      9          Shane Acker   
3      4     21       Robert Luketic   
4      6    122      Yasir Al Yasiri   
5      7    187       Kevin Reynolds   
6      8    706        Shravan Kumar   
7      9   1920         Vikram Bhatt   
8     10   1922         Zak Hilditch   
9     13  2,215  Nottapon Boonprakob   

                                                cast        country  \
0  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
1  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
2  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
3  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   
4  Amina Khalil, Ahmed Dawood, Tarek Lotfy, Ahmed...          Egypt   
5  Samuel L. Jackson, John Heard, Kelly Rowan, Cl...  United States   
6  Divya Dutta, Atul Kulkarni, Mohan Agashe, Anup...          India   
7  Rajneesh Duggal, Adah Sharma, Indraneil Sengup...          India   
8  Thomas Jane, Molly Parker, Dylan Schmid, Kaitl...  United States   
9                                 Artiwara Kongmalai       Thailand   

          date_added  release_year rating duration  \
0  December 23, 2016          2016  TV-MA   93 min   
1  December 20, 2018          2011      R   78 min   
2  November 16, 2017          2009  PG-13   80 min   
3    January 1, 2020          2008  PG-13  123 min   
4       June 1, 2020          2019  TV-MA   95 min   
5   November 1, 2019          1997      R  119 min   
6      April 1, 2019          2019  TV-14  118 min   
7  December 15, 2017          2008  TV-MA  143 min   
8   October 20, 2017          2017  TV-MA  103 min   
9      March 1, 2019          2018  TV-MA   89 min   

                                           listed_in  \
0                       Dramas, International Movies   
1                Horror Movies, International Movies   
2  Action & Adventure, Independent Movies, Sci-Fi...   
3                                             Dramas   
4                Horror Movies, International Movies   
5                                             Dramas   
6                Horror Movies, International Movies   
7     Horror Movies, International Movies, Thrillers   
8                                  Dramas, Thrillers   
9  Documentaries, International Movies, Sports Mo...   

                                         description  movie_id  
0  After a devastating earthquake hits Mexico Cit...       NaN  
1  When an army recruit is found dead, his fellow...       NaN  
2  In a postapocalyptic world, rag-doll robots hi...       NaN  
3  A brilliant group of students become card-coun...       NaN  
4  After an awful accident, a couple admitted to ...       NaN  
5  After one of his high school students attacks ...       NaN  
6  When a doctor goes missing, his psychiatrist w...       NaN  
7  An architect and his wife move into a castle t...       NaN  
8  A farmer pens a confession admitting to his wi...       NaN  
9  This intimate documentary follows rock star Ar...       NaN

In [18]:
movie_df['movie_id'] = movie_df.apply(lambda x: get_movie_id(x['title']), axis=1)

In [19]:
movie_df

index                                    title           director  \
0         1                                     7:19  Jorge Michel Grau   
1         2                                    23:59       Gilbert Chan   
2         3                                        9        Shane Acker   
3         4                                       21     Robert Luketic   
4         6                                      122    Yasir Al Yasiri   
...     ...                                      ...                ...   
5372   7781                                     Zoom       Peter Hewitt   
5373   7782                                     Zozo        Josef Fares   
5374   7783                                   Zubaan        Mozez Singh   
5375   7784                        Zulu Man in Japan                NaN   
5376   7786  ZZ TOP: THAT LITTLE OL' BAND FROM TEXAS           Sam Dunn   

                                                   cast  \
0     Demián Bichir, Héctor Bonilla, Oscar Serrano, ...   
1     Tedd Chan, Stella Chung, Henley Hii, Lawrence ...   
2     Elijah Wood, John C. Reilly, Jennifer Connelly...   
3     Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...   
4     Amina Khalil, Ahmed Dawood, Tarek Lotfy, Ahmed...   
...                                                 ...   
5372  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
5373  Imad Creidi, Antoinette Turk, Elias Gergi, Car...   
5374  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   
5375                                            Nasty C   
5376                                                NaN   

                                                country          date_added  \
0                                                Mexico   December 23, 2016   
1                                             Singapore   December 20, 2018   
2                                         United States   November 16, 2017   
3                                         United States     January 1, 2020   
4                                                 Egypt        June 1, 2020   
...                                                 ...                 ...   
5372                                      United States    January 11, 2020   
5373  Sweden, Czech Republic, United Kingdom, Denmar...    October 19, 2020   
5374                                              India       March 2, 2019   
5375                                                NaN  September 25, 2020   
5376              United Kingdom, Canada, United States       March 1, 2020   

      release_year rating duration  \
0             2016  TV-MA   93 min   
1             2011      R   78 min   
2             2009  PG-13   80 min   
3             2008  PG-13  123 min   
4             2019  TV-MA   95 min   
...            ...    ...      ...   
5372          2006     PG   88 min   
5373          2005  TV-MA   99 min   
5374          2015  TV-14  111 min   
5375          2019  TV-MA   44 min   
5376          2019  TV-MA   90 min   

                                              listed_in  \
0                          Dramas, International Movies   
1                   Horror Movies, International Movies   
2     Action & Adventure, Independent Movies, Sci-Fi...   
3                                                Dramas   
4                   Horror Movies, International Movies   
...                                                 ...   
5372                 Children & Family Movies, Comedies   
5373                       Dramas, International Movies   
5374     Dramas, International Movies, Music & Musicals   
5375  Documentaries, International Movies, Music & M...   
5376                    Documentaries, Music & Musicals   

                                            description  movie_id  
0     After a devastating earthquake hits Mexico Cit...  387805.0  
1     When an army recruit is found dead, his fellow...  107227.0  
2     In a postapocalyptic world, rag-doll robots hi...   12244.0  
3     A b

<p>Now we have filled up most <b>movie_id</b>, we should be able to easily retrieve the other data.</p>
<p>Before that, we should also save the file as a new checkpoint</p>

In [21]:
# movie_df.to_csv('movie_df.csv', index = False)

In [389]:
movie_df = pd.read_csv('movie_df.csv')

In [390]:
movie_df.isnull().sum()

index             0
title             0
director        163
cast            426
country         230
date_added        0
release_year      0
rating            5
duration          0
listed_in         0
description       0
movie_id        207
dtype: int64

In [391]:
def find_director(row):
    movie_id = int(row['movie_id'])   
    response = requests.get(
    'https://api.themoviedb.org/3/movie/{}/credits'.format(movie_id),
    params={'api_key': TMDB_API,},
    )

    director = ''
    for crew in response.json()['crew']:
        if crew['job'] == 'Director':
            director += crew['original_name'] + ', '
#         print(director)
    return str(director[:-1])


def find_cast(row):
    movie_id = int(row['movie_id'])
    response = requests.get(
    'https://api.themoviedb.org/3/movie/{}/credits'.format(movie_id),
    params={'api_key': TMDB_API,},
    )

    cast_list = ''
    for cast in response.json()['cast']:
        cast_list += cast['name'] + ', '
#     print(cast_list)
    return str(cast_list[:-1])


def find_country(row):
    movie_id = int(row['movie_id'])


#             print('finding country')
    response = requests.get(
    'https://api.themoviedb.org/3/movie/{}'.format(movie_id),
    params={'api_key': TMDB_API,},
    )

#     print(response.status_code)
    try:
        country_list = ''
        for country in response.json()['production_countries']:
            country_list += country['name'] + ', '
        
#         country_list = response.json()['production_countries'][0]['name']
#         print('country list: ', country_list)
        return str(country_list[:-1])
    except:
        return(row['country'])





In [392]:
missing_data = movie_df[movie_df[['country','director','cast']].isnull().any(axis=1)]


In [393]:
# drop missing data with no movie_id

missing_data = missing_data.dropna(subset=['movie_id']).copy()

In [394]:
missing_data['country'] = missing_data.apply(lambda x: find_country(x), axis=1)

In [395]:
missing_data['director'] = missing_data.apply(lambda x: find_director(x), axis=1)

In [396]:
missing_data['cast'] = missing_data.apply(lambda x: find_cast(x), axis=1)

In [397]:
missing_data.isnull().sum()

index           0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          1
duration        0
listed_in       0
description     0
movie_id        0
dtype: int64

In [398]:
missing_data.drop(columns ='index', inplace = True)

In [399]:
missing_data

title  \
14                                        '89   
19                                   (T)ERROR   
22                           #cats_the_mewvie   
27                                  #Rucker50   
41                       100 Days Of Solitude   
...                                       ...   
5332      Yoo Byung Jae: Too Much Information   
5345                          Your Excellency   
5361                                Zero Hour   
5375                        Zulu Man in Japan   
5376  ZZ TOP: THAT LITTLE OL' BAND FROM TEXAS   

                                     director  \
14                              Dave Stewart,   
19    Lyric R. Cabral, David Felix Sutcliffe,   
22                          Michael Margolis,   
27                     Robert McCullough Jr.,   
41      Gerardo Olivares, José Díaz Martínez,   
...                                       ...   
5332                           Yoo Byung-jae,   
5345                          Funke Akindele,   
5361                           Hall Bartlett,   
5375                                            
5376                                Sam Dunn,   

                                                   cast  \
14    Lee Dixon, Ian Wright, Paul Merson, Steve Boul...   
19                                                        
22    Mike Bridavsky, Jason Eppink, Amanda Brennan, ...   
27                                                        
41                                           José Díaz,   
...                                                 ...   
5332                      Yoo Byung-jae, Yang Hyun-suk,   
5345  Akin Lewis, Funke Akindele, Kemi Lala Akindoju...   
5361  Dana Andrews, Linda Darnell, Sterling Hayden, ...   
5375                                           Nasty C,   
5376  Frank Beard, Billy Gibbons, Dusty Hill, Joshua...   

                        country          date_added  release_year rating  \
14              United Kingdom,        May 16, 2018          2017  TV-PG   
19    United States of America,       June 30, 2016          2015     NR   
22                      Canada,    February 5, 2020          2020  TV-14   
27                                 December 1, 2016          2016  TV-PG   
41                       Spain,        July 6, 2018          2018  TV-MA   
...                         ...                 ...           ...    ...   
5332               South Korea,      March 16, 2018          2018  TV-MA   
5345                   Nigeria,       July 10, 2020          2019   TV-G   
5361  United States of America,   December 13, 2019          2018  TV-MA   
5375              South Africa,  September 25, 2020          2019  TV-MA   
5376            United Kingdom,       March 1, 2020          2019  TV-MA   

     duration                                          listed_in  \
14     87 min                                      Sports Movies   
19     84 min                                      Documentaries   
22     90 min                Documentaries, International Movies   
27     56 min                       Documentaries, Sports Movies   
41     93 min                Documentaries, International Movies   
...       ...                                                ...   
5332   63 min                                    Stand-Up Comedy   
5345  120 min                     Comedies, International Movies   
5361   89 min                    International Movies, Thrillers   
5375   44 min  Documentaries, International Movies, Music & M...   
5376   90 min                    Documentaries, Music & Musicals   

                                            description  movie_id  
14    Mixing old footage with interviews, this is th...  486882.0  
19    This real-life look at FBI counterterrorism op...  318972.0  
22    This pawesome documentary explores how our fel...  668291.0  
27    This documentary celebrates the 50th anniversa...  433479.0  
41    Spanish photographer José Díaz spends 100 days...  534110.0  
...            

<p>Now we should drop the data from original <b>movie_df</b> and merge the processed (missing_data) dataframe back into the original movie_df</p>

In [400]:
movie_df = movie_df.drop(missing_data.index[:])

In [401]:
movie_df = (movie_df.append(missing_data))

In [402]:
movie_df.reset_index(inplace=True)

In [403]:
movie_df.drop(columns=['level_0', 'index'], inplace = True)

<h4>Dropping some rows with missing data</h4>
<p>We have attempted to fill in most of the data, and let's see if we have reduced the occurence of missing data.</p>

In [404]:
print(movie_df.isnull().sum())
after = movie_df.isnull().sum()

title             0
director         36
cast             25
country          58
date_added        0
release_year      0
rating            5
duration          0
listed_in         0
description       0
movie_id        207
dtype: int64


<p>Let's see how well did we do.</p>

In [405]:
print(before)

show_id           0
title             0
director        163
cast            426
country         230
date_added        0
release_year      0
rating            5
duration          0
listed_in         0
description       0
dtype: int64


<p>Sadly, there are still some missing data. So I guess we would just drop them and movie_id column for now for the sake of analysis. After that, let's save the dataframe into another checkpoint just in case.</p>

In [406]:
movie_df.drop(columns=['movie_id'], axis = 1, inplace=True)
movie_df.dropna(axis=0, how='any', inplace=True)

<p>Ok, let's check if everything looks good before storing the dataframe into a new checkpoint file.</p>

In [408]:
movie_df.head(5)

title           director  \
0   7:19  Jorge Michel Grau   
1  23:59       Gilbert Chan   
2      9        Shane Acker   
3     21     Robert Luketic   
4    122    Yasir Al Yasiri   

                                                cast        country  \
0  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
1  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
2  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
3  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   
4  Amina Khalil, Ahmed Dawood, Tarek Lotfy, Ahmed...          Egypt   

          date_added  release_year rating duration  \
0  December 23, 2016          2016  TV-MA   93 min   
1  December 20, 2018          2011      R   78 min   
2  November 16, 2017          2009  PG-13   80 min   
3    January 1, 2020          2008  PG-13  123 min   
4       June 1, 2020          2019  TV-MA   95 min   

                                           listed_in  \
0                       Dramas, International Movies   
1                Horror Movies, International Movies   
2  Action & Adventure, Independent Movies, Sci-Fi...   
3                                             Dramas   
4                Horror Movies, International Movies   

                                         description  
0  After a devastating earthquake hits Mexico Cit...  
1  When an army recruit is found dead, his fellow...  
2  In a postapocalyptic world, rag-doll robots hi...  
3  A brilliant group of students become card-coun...  
4  After an awful accident, a couple admitted to ...

In [419]:
movie_df.notnull().sum()

title           5042
director        5042
cast            5042
country         5042
date_added      5042
release_year    5042
rating          5042
duration        5042
listed_in       5042
description     5042
dtype: int64

In [410]:
movie_df.to_csv('movie_df_cp2.csv', index = False)

In [512]:
movie_df = pd.read_csv('movie_df_cp2.csv')

In [513]:
movie_df.dropna(axis=0, how='any', inplace=True)

<p>Now that we have filled missing data, which by the way, as listed below</p> 

<table>
    <tr>
        <th>feature</th>
        <th>before</th>
        <th>after</th>
        <th>leftover</th>
    </tr>
    <tr>
        <td>director</td>
        <td>163</td>
        <td>36</td>
        <td>5042</td>
    </tr>  
    <tr>
        <td>cast</td>
        <td>426</td>
        <td>25</td>
        <td>5042</td>
    </tr>
    <tr>
        <td>country</td>
        <td>230</td>
        <td>58</td>
        <td>5042</td>
    </tr>
</table>

<p>We should look into the data again, and as we could see, we could still do a bit of preprocessing first.</p> 


In [514]:
movie_df.head(4)

title           director  \
0   7:19  Jorge Michel Grau   
1  23:59       Gilbert Chan   
2      9        Shane Acker   
3     21     Robert Luketic   

                                                cast        country  \
0  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
1  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
2  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
3  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   

          date_added  release_year rating duration  \
0  December 23, 2016          2016  TV-MA   93 min   
1  December 20, 2018          2011      R   78 min   
2  November 16, 2017          2009  PG-13   80 min   
3    January 1, 2020          2008  PG-13  123 min   

                                           listed_in  \
0                       Dramas, International Movies   
1                Horror Movies, International Movies   
2  Action & Adventure, Independent Movies, Sci-Fi...   
3                                             Dramas   

                                         description  
0  After a devastating earthquake hits Mexico Cit...  
1  When an army recruit is found dead, his fellow...  
2  In a postapocalyptic world, rag-doll robots hi...  
3  A brilliant group of students become card-coun...

<p>We could start by removing the "min" in the duration column, which could be done by a simple function.</p>

In [515]:
movie_df.iloc[0]['duration']

'93 min'

In [516]:
def remove_unit(input_value, unit_size):
    value = input_value[:-unit_size]
    return value

In [517]:
movie_df['duration'] = movie_df.apply(lambda x: remove_unit(x['duration'], 4), axis=1)

In [518]:
movie_df.head(3)

title           director  \
0   7:19  Jorge Michel Grau   
1  23:59       Gilbert Chan   
2      9        Shane Acker   

                                                cast        country  \
0  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
1  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
2  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   

          date_added  release_year rating duration  \
0  December 23, 2016          2016  TV-MA       93   
1  December 20, 2018          2011      R       78   
2  November 16, 2017          2009  PG-13       80   

                                           listed_in  \
0                       Dramas, International Movies   
1                Horror Movies, International Movies   
2  Action & Adventure, Independent Movies, Sci-Fi...   

                                         description  
0  After a devastating earthquake hits Mexico Cit...  
1  When an army recruit is found dead, his fellow...  
2  In a postapocalyptic world, rag-doll robots hi...

<p>Also we should check if the country names have been consistent, for example I noticed that USA was named "United States" in the original data, but "United States of America" in the TMDB data. So we should check if we need to conduct a bit of preprocessing there.</p>

In [519]:
def replace_country_name(country_list, find_word, replaced_by):
    return country_list.replace(find_word, replaced_by).rstrip(',')

In [520]:
movie_df['country'].value_counts()

United States                1627
India                         827
United States of America,     179
United Kingdom                155
Canada                        101
                             ... 
Singapore, Malaysia             1
United Kingdom, Belgium         1
United States, Brazil           1
Uruguay, Guatemala              1
Nigeria, United Kingdom         1
Name: country, Length: 619, dtype: int64

In [521]:
movie_df['country'] = movie_df.apply(lambda x: replace_country_name(x['country'], 'United States of America', 'United States'), axis=1)

In [522]:
movie_df['country'].value_counts()

United States                                                   1806
India                                                            841
United Kingdom                                                   181
Canada                                                           110
Egypt                                                             96
                                                                ... 
Sweden, Czech Republic, United Kingdom, Denmark, Netherlands       1
United States, Hungary                                             1
Singapore, Malaysia                                                1
United Kingdom, Belgium                                            1
Nigeria, United Kingdom                                            1
Name: country, Length: 567, dtype: int64

In [523]:
country_df = movie_df['country'].str.get_dummies(', ')

In [525]:
combined_movie_df = pd.concat([movie_df, country_df], axis=1)

In [526]:
combined_movie_df.drop(columns = 'country', axis = 1, inplace = True)

In [527]:
combined_movie_df.head(5)

title           director  \
0   7:19  Jorge Michel Grau   
1  23:59       Gilbert Chan   
2      9        Shane Acker   
3     21     Robert Luketic   
4    122    Yasir Al Yasiri   

                                                cast         date_added  \
0  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...  December 23, 2016   
1  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...  December 20, 2018   
2  Elijah Wood, John C. Reilly, Jennifer Connelly...  November 16, 2017   
3  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...    January 1, 2020   
4  Amina Khalil, Ahmed Dawood, Tarek Lotfy, Ahmed...       June 1, 2020   

   release_year rating duration  \
0          2016  TV-MA       93   
1          2011      R       78   
2          2009  PG-13       80   
3          2008  PG-13      123   
4          2019  TV-MA       95   

                                           listed_in  \
0                       Dramas, International Movies   
1                Horror Movies, International Movies   
2  Action & Adventure, Independent Movies, Sci-Fi...   
3                                             Dramas   
4                Horror Movies, International Movies   

                                         description  Afghanistan  ...  \
0  After a devastating earthquake hits Mexico Cit...            0  ...   
1  When an army recruit is found dead, his fellow...            0  ...   
2  In a postapocalyptic world, rag-doll robots hi...            0  ...   
3  A brilliant group of students become card-coun...            0  ...   
4  After an awful accident, a couple admitted to ...            0  ...   

   Ukraine  United Arab Emirates  United Kingdom  United States  Uruguay  \
0        0                     0               0              0        0   
1        0                     0               0              0        0   
2        0                     0               0              1        0   
3        0                     0               0              1        0   
4        0                     0               0              0        0   

   Vatican City  Venezuela  Vietnam  West Germany  Zimbabwe  
0             0          0        0             0         0  
1             0          0        0             0         0  
2             0          0        0             0         0  
3             0          0        0             0         0  
4             0          0        0             0         0  

[5 rows x 118 columns]

<p>Same goes for the cast, listed_in and director columns. </p>

In [528]:
director_df = movie_df['director'].str.get_dummies(', ')
cast_df = movie_df['cast'].str.get_dummies(', ')
genre_df = movie_df['listed_in'].str.get_dummies(', ')

combined_movie_df = pd.concat([combined_movie_df, director_df], axis=1)
combined_movie_df = pd.concat([combined_movie_df, cast_df], axis=1)
combined_movie_df = pd.concat([combined_movie_df, genre_df], axis=1)

combined_movie_df.drop(columns = ['director', 'cast', 'listed_in'])

title         date_added  \
0                                        7:19  December 23, 2016   
1                                       23:59  December 20, 2018   
2                                           9  November 16, 2017   
3                                          21    January 1, 2020   
4                                         122       June 1, 2020   
...                                       ...                ...   
5279                     Yoko and His Friends      June 23, 2018   
5280      Yoo Byung Jae: Too Much Information     March 16, 2018   
5281                          Your Excellency      July 10, 2020   
5282                                Zero Hour  December 13, 2019   
5284  ZZ TOP: THAT LITTLE OL' BAND FROM TEXAS      March 1, 2020   

      release_year rating duration  \
0             2016  TV-MA       93   
1             2011      R       78   
2             2009  PG-13       80   
3             2008  PG-13      123   
4             2019  TV-MA       95   
...            ...    ...      ...   
5279          2015   TV-Y       78   
5280          2018  TV-MA       63   
5281          2019   TV-G      120   
5282          2018  TV-MA       89   
5284          2019  TV-MA       90   

                                            description  Afghanistan  Albania  \
0     After a devastating earthquake hits Mexico Cit...            0        0   
1     When an army recruit is found dead, his fellow...            0        0   
2     In a postapocalyptic world, rag-doll robots hi...            0        0   
3     A brilliant group of students become card-coun...            0        0   
4     After an awful accident, a couple admitted to ...            0        0   
...                                                 ...          ...      ...   
5279  Vik meets new friends in a new city, where the...            0        0   
5280  "Saturday Night Live Korea" writer-turned-come...            0        0   
5281  Bumbling through politics, a billionaire busin...            0        0   
5282  After his father passes, the heir to a retail ...            0        0   
5284  This documentary delves into the mystique behi...            0        0   

      Algeria  Angola  ...  Independent Movies  International Movies  \
0           0       0  ...                   0                     1   
1           0       0  ...                   0                     1   
2           0       0  ...                   1                     0   
3           0       0  ...                   0                     0   
4           0       0  ...                   0                     1   
...       ...     ...  ...                 ...                   ...   
5279        0       0  ...                   0                     0   
5280        0       0  ...                   0                     0   
5281        0       0  ...                   0                     1   
5282        0       0  ...                   0                     1   
5284        0       0  ...                   0                     0   

      LGBTQ Movies  Movies  Music & Musicals  Romantic Movies  \
0                0       0                 0                0   
1                0       0                 0                0   
2                0       0                 0                0   
3                0       0                 0                0   
4                0       0                 0                0   
...            ...     ...               ...              ...   
5279             0       0                 0                0   
5280             0       0                 0                0   
5281             0       0                 0                0   
5282             0       0                 0                0   
5284             0       0                 1                0   

      Sci-Fi & Fantasy  Sports Movies  Stand-Up Comedy  Thrillers  
0                    0              0                0          0  
1          

<h4>Exploratory Data Analysis</h4>
<p>The next step is to perform exploratory data analysis on the dataset. We can refer to <a href="https://www.datacourses.com/tips-for-performing-eda-with-python-671/">this page</a>.</p>

In [530]:
combined_movie_df.drop(columns = ['director', 'cast', 'listed_in', 'description'], inplace = True)

In [538]:
combined_movie_df.tail(3)

title         date_added  \
5281                          Your Excellency      July 10, 2020   
5282                                Zero Hour  December 13, 2019   
5284  ZZ TOP: THAT LITTLE OL' BAND FROM TEXAS      March 1, 2020   

      release_year rating duration  Afghanistan  Albania  Algeria  Angola  \
5281          2019   TV-G      120            0        0        0       0   
5282          2018  TV-MA       89            0        0        0       0   
5284          2019  TV-MA       90            0        0        0       0   

      Argentina  ...  Independent Movies  International Movies  LGBTQ Movies  \
5281          0  ...                   0                     1             0   
5282          0  ...                   0                     1             0   
5284          0  ...                   0                     0             0   

      Movies  Music & Musicals  Romantic Movies  Sci-Fi & Fantasy  \
5281       0                 0                0                 0   
5282       0                 0                0                 0   
5284       0                 1                0                 0   

      Sports Movies  Stand-Up Comedy  Thrillers  
5281              0                0          0  
5282              0                0          1  
5284              0                0          0  

[3 rows x 30221 columns]

In [539]:
combined_movie_df.to_csv('combined_movie_df.csv', index = False)